src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/boss_skadi.cpp

```diff
@@ -270,196 +215,433 @@ class boss_skadi : public CreatureScript
             {
-                case 0:
-                    iStart = 8;
-                    iEnd = 37;
-                    break;
-                case 1:
-                    iStart = 38;
-                    iEnd = 68;
+                case EVENT_MOVEMENT:
+                    switch (step)
+                    {
+                        case 0:
+                            me->GetMotionMaster()->MovePoint(0, Location[2]);
+                            events.ScheduleEvent(EVENT_MOVEMENT, 2000);
+                            break;
+                        case 1:
+                            me->GetMotionMaster()->MovePoint(0, Location[3]);
+                            events.ScheduleEvent(EVENT_EMOTE, 12500);
+                            events.ScheduleEvent(EVENT_MOVEMENT, 22500);
+                            break;
+                        case 2:
+                            me->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NOT_SELECTABLE | UNIT_FLAG_NON_ATTACKABLE);
+                            me->GetMotionMaster()->MovePoint(0, Location[6]);
+                            events.ScheduleEvent(EVENT_EMOTE, 2000);
+                            events.ScheduleEvent(EVENT_MOVEMENT, 5000);
+                            break;
+                        case 3:
+                            if (Creature* skadi = ObjectAccessor::GetCreature(*me, skadiGUID))
+                                skadi->AI()->DoAction(DATA_FLAME);
+                            DoCast(RAND(SPELL_FREEZING_CLOUD_LEFT_PERIODIC, SPELL_FREEZING_CLOUD_RIGHT_PERIODIC));
+                            me->GetMotionMaster()->MovePoint(0, Location[7]);
+                            events.ScheduleEvent(EVENT_MOVEMENT, 10000);
+                            break;
+                        case 4:
+                            me->RemoveAurasDueToSpell(SPELL_FREEZING_CLOUD_LEFT_PERIODIC);
+                            me->RemoveAurasDueToSpell(SPELL_FREEZING_CLOUD_RIGHT_PERIODIC);
+                            me->GetMotionMaster()->MovePoint(0, Location[1]);
+                            events.ScheduleEvent(EVENT_MOVEMENT, 3500);
+                            break;
+                    }
+                    step++;
+                    if (step == 5)
+                        step = 0;
                     break;
-                default:
+                case EVENT_EMOTE:
+                    switch (step)
+                    {
+                        case 2:
+                            me->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NOT_SELECTABLE | UNIT_FLAG_NON_ATTACKABLE);
+                            me->GetMotionMaster()->MovePoint(0, Location[9]);
+                            if (!triggers.empty())
+                            {
+                                std::list<Creature*>::const_iterator itr = triggers.begin();
+                                for (itr; itr != triggers.end(); ++itr)
+                                    (*itr)->AI()->SetData(0, 0);
```

Comment:
Same as the comment above, are you sure it's safe to store a Creature* and dereference it ?